framework/build.gradle

```diff
@@ -50,8 +50,8 @@ android {
     buildToolsVersion cordovaConfig.BUILD_TOOLS_VERSION
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_17
```

Comment:
Hey guys, quick question I am struggling to find answers for:
Considering the `minSdk` currently in effect (API Level 24), how can you be sure there won't be any issues targeting Java 17?

As of today, generating a pure Android project using AS Hedgehog RC and AGP 8.2.0-rc03 still results in a Gradle file containing:
```
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
```
It seems Google is not ready to pull the trigger on Java 17 yet, so I am wondering if Cordova should?

Reply:
JDK 17 compatibility was [added in AGP 8.1](https://developer.android.com/build/releases/gradle-plugin#8-1-0)

> how can you be sure there won't be any issues targeting Java 17?

It's my understanding that it will enable JDK 17 features on supported devices. Different devices will have different java runtimes,so plugin authors may still need to do an API check. But targeting JDK 17 doesn't necessarily limit the ability to run on older devices.

Resources:
https://developer.android.com/build/jdks#compileSdk
https://developer.android.com/about/versions/14/features#core-libraries

Though you do bring a good point, it's hard to say with any level of certainty that this won't break numerous of plugins that are likely built with JDK 8 as a target. If there are any breaking changes in the language, the compilation target may break builds. It hasn't in my own testing with the plugins I have used, it might be smart to introduce a preference to set these fields.

Reply:
> JDK 17 compatibility was [added in AGP 8.1](https://developer.android.com/build/releases/gradle-plugin#8-1-0)

That is a requirement to run AGP itself and does not seem to have anything to do with how the project is setup with regards to JVM targeting.

> Resources:
https://developer.android.com/build/jdks#compileSdk
https://developer.android.com/about/versions/14/features#core-libraries

From my understanding, the documentation refers to applications built with JDK 17 as a target + running on Android 14.

For 13 and below it would require desugaring of JDK 17 APIs via the `com.android.tools:desugar_jdk_libs:2.0.3` dependency as per https://developer.android.com/studio/write/java8-support#library-desugaring.

It is hard to tell because that documentation seems stopped in time (unlike the dependency itself which was updated [earlier this month](https://mvnrepository.com/artifact/com.android.tools/desugar_jdk_libs/2.0.4)), with no mention to AGP 8 or JDK 17 for that matter. The [changelog](https://github.com/google/desugar_jdk_libs/blob/master/CHANGELOG.md) of the dependency does not mention JDK 17 either.

The way I see it if you want to perform this change you would need to add the dependency and set the `isCoreLibraryDesugaringEnabled` property to `true`, assuming that it is indeed capable to desugar JDK 17 APIs.

What are your thoughts on this?

Reply:
I've revert the change back to 8, but powered the source and target compatibility options via 2 new preferences so that it can be experimented with. The default will remain using Java 8.

Overall I do agree that we should try to maintain the Android Studio default as much as possible.

Reply:
Added `AndroidKotlinJVMTarget`, which will only be effective if kotlin is enabled (by default its disabled but there is an existing preference to enable it). The default value is for the JVM target `1.8`.

Unlike the java source / target compatibility, `AndroidKotlinJVMTarget` only applies to the app target, not the cordova framework. Cordova framework currently doesn't setup/enable kotlin nor does it use kotlin, so it doesn't read or set the JVM target. Trying to do so will cause build errors.